cloudflare-workers/src/db.ts

```diff
@@ -20,18 +20,41 @@ export async function postComment(env: Env, req: PostComment) {
 		offset = (await db.prepare('SELECT last_insert_rowid() AS id').first())!;
 	}
 
-	await db
-		.prepare('INSERT INTO commenters (name, user_agent, ip_address) VALUES (?, ?, ?)')
-		.bind(req.commenter?.name, req.commenter?.user_agent, req.commenter?.ip_address)
-		.run();
-	const commiterId = (await db.prepare('SELECT last_insert_rowid() AS id').first())!.id;
+	const user = await db
+		.prepare('SELECT id FROM commenters WHERE oauth_provider = ? AND oauth_user_id = ?')
+		.bind(req.commenter.oauth_provider, req.commenter.oauth_user_id)
+		.first();
+	if (!user) {
+		throw new Error('User not found, please try to re-login.');
+	}
+	const commiterId = user.id;
 
 	await db
 		.prepare('INSERT INTO comments (offset_id, commenter_id, comment, created_time) VALUES (?, ?, ?, ?)')
 		.bind(offset.id, commiterId, req.comment, new Date().toISOString())
 		.run();
 }
 
+export async function registerUser(env: Env, name: string, oauth_provider: string, oauth_user_id: string) {
+	const db = env.DB;
+
+	const user = await db
+		.prepare('SELECT id, name FROM commenters WHERE oauth_provider = ? AND oauth_user_id = ?')
+		.bind(oauth_provider, oauth_user_id)
+		.first();
+	if (user) {
+		if (user.name !== name) {
+			await db.prepare('UPDATE commenters SET name = ? WHERE id = ?').bind(name, user.id).run();
+		}
+		return;
+	}
+
+	await db
+		.prepare('INSERT INTO commenters (name, oauth_provider, oauth_user_id) VALUES (?, ?, ?)')
+		.bind(name, oauth_provider, oauth_user_id)
+		.run();
+}
+
```

Comment:
这一坨应该也可以搞一下，大概是这样（GPT教我的魔法，感觉意外的还没几行就实现了）

```
        INSERT OR REPLACE INTO commenters (id, name, oauth_provider, oauth_user_id)
        VALUES (
            (SELECT id FROM commenters WHERE oauth_provider = ? AND oauth_user_id = ?),
            ?,
            ?,
            ?
        )
```

Reply:
还有这种行为，这个确实不太清楚。。那用 on conflict 可以吗

```
INSERT INTO commenters (name, oauth_provider, oauth_user_id)
VALUES (?, ?, ?)
ON CONFLICT(oauth_provider, oauth_user_id) 
DO UPDATE SET name = excluded.name;
```

如果不冲突就正常插入，否则更新名字

Reply:
草，ON CONFILICT REPLACE 其实就是 INSERT OR REPLACE
http://sqlite.org/lang_conflict.html
> The syntax for the ON CONFLICT clause is as shown above for the CREATE TABLE command. For the INSERT and UPDATE commands, the keywords "ON CONFLICT" are replaced by "OR" so that the syntax reads more naturally. For example, instead of "INSERT ON CONFLICT IGNORE" we have "INSERT OR IGNORE". The keywords change but the meaning of the clause is the same either way.

Reply:
我发的写法看起来实际上是 https://sqlite.org/lang_upsert.html 这个文档

可以在本地跑这个可以发现它是原地改的（而不是删了重插）
```sql
-- 创建表
CREATE TABLE IF NOT EXISTS `commenters` (
    `id` INTEGER PRIMARY KEY AUTOINCREMENT,
    `oauth_provider` TEXT NOT NULL,
    `oauth_user_id` TEXT NOT NULL,
    `name` TEXT NOT NULL
);

-- 创建复合唯一索引
CREATE UNIQUE INDEX IF NOT EXISTS idx_oauth_provider_user_id 
ON commenters(oauth_provider, oauth_user_id);

-- 插入一些初始数据
INSERT INTO commenters (oauth_provider, oauth_user_id, name) VALUES
('google', '12345', 'Alice'),
('facebook', '67890', 'Bob'),
('twitter', '11111', 'Charlie');

-- 显示初始数据
SELECT * FROM commenters;

-- 尝试插入一个新记录
INSERT INTO commenters (oauth_provider, oauth_user_id, name)
VALUES ('github', '22222', 'David')
ON CONFLICT(oauth_provider, oauth_user_id) 
DO UPDATE SET name = excluded.name;

-- 尝试更新一个现有记录
INSERT INTO commenters (oauth_provider, oauth_user_id, name)
VALUES ('google', '12345', 'Alice Updated')
ON CONFLICT(oauth_provider, oauth_user_id) 
DO UPDATE SET name = excluded.name;

-- 显示最终结果
SELECT * FROM commenters;

```